builder/dockerfile/evaluator.go

```diff
@@ -192,12 +111,30 @@ type dispatchState struct {
 	imageID    string
 	baseImage  builder.Image
 	stageName  string
+	buildArgs  *buildArgs
 }
 
-func newDispatchState() *dispatchState {
-	return &dispatchState{runConfig: &container.Config{}}
+type stageBuild struct {
+	state   *dispatchState
+	shlex   *ShellLex
+	builder *Builder
+	source  builder.Source
+}
+
+func newDispatchState(baseArgs *buildArgs) *dispatchState {
+	args := baseArgs.Clone()
+	args.ResetAllowed()
+	return &dispatchState{runConfig: &container.Config{}, buildArgs: args}
 }
 
+func newStageBuild(builder *Builder, escapeToken rune, source builder.Source, buildArgs *buildArgs) *stageBuild {
```

Comment:
Minor nit. The definition for `dispatchState` and `StageBuild` are a bit tangled here. Could you move this above `newDispatchState()`, so their "constructors" are together with the struct.